File: smart_solutions/smart_solutions/report/drivers_and_authorized_vehicles/drivers_and_authorized_vehicles.js
// Copyright (c) 2025, Asofi and contributors
// For license information, please see license.txt

frappe.query_reports["Drivers and Authorized Vehicles"] = {
    tree: true,
    name_field: "driver_or_vehicle",
    parent_field: "parent",
    initial_depth: 2,

    formatter: function(value, row, column, data, default_formatter) {
        let v = default_formatter(value, row, column, data);

        const hasValue = v && v !== "null" && v !== "undefined" && v.trim() !== "";
        if (!hasValue) return "";

        if (data.is_group && column.fieldname === "driver_or_vehicle") {
            const iconClass = data.expanded ? "expanded" : "collapsed";
            const toggle = `<span class="tree-toggle ${iconClass}">▶</span>`;
            v = toggle + " " + v;
        }

        if (data.is_group) {
            v = `<span style="background: #5cb85c; color: white; padding: 3px 6px; border-radius: 3px;">${v}</span>`;
        } else if (data.indent) {
            v = `<span style="background: #d9edf7; color: #000; padding: 3px 6px; border-radius: 3px;">${v}</span>`;
        }

        return v;
    },

    filters: [
        {
            fieldname: "driver",
            label: __("Driver"),
            fieldtype: "Link",
            options: "Driver",
        },
    ]
};

File: smart_solutions/smart_solutions/report/drivers_and_authorized_vehicles/drivers_and_authorized_vehicles.py
# Copyright (c) 2025, Asofi and contributors
# For license information, please see license.txt

import frappe

def execute(filters=None):
    columns = [
        {"fieldname": "driver", "label": "Name", "fieldtype": "Data", "width": 300},
        {"fieldname": "vehicle", "label": "Vehicle", "fieldtype": "Data", "width": 300},
        {"fieldname": "make", "label": "Make", "fieldtype": "Data", "width": 200},
        {"fieldname": "model", "label": "Model", "fieldtype": "Data", "width": 300},
    ]
    data = []

    drivers = frappe.get_all("Driver", fields=["name"])
    for d in drivers:
        data.append({
            "driver": d.name,
            "parent": None,
            "is_group": True,
            "indent": 0
        })
        rows = frappe.get_all("Authorized Vehicle", filters={"parent": d.name}, fields=["vehicle","model","make"])
        for r in rows:
            v = frappe.get_doc("Vehicle", r.vehicle)
            data.append({
                "vehicle": v.get('license_plate') or v.name,
                "model": v.get('model'),
                "make": v.get('make'),
                "parent": d.name,
                "is_group": False,
                "indent": 1
            })

    return columns, data

